### PR TITLE
Dynamic changes for AndesButton properties

### PR DESCRIPTION
**Pull request by lmadrazo**
## Motivation
AndesButton hierarchy could not be updated after its creation. This PR includes changes to support dynamically change its configuration.

## Description
- What is it about?
AndesButton configuration can be updated in runtime, consistently with AndesMessage
(see #7).
- How does it work?
This PR adds the possibility to update the button hierarchy or text after its creation. See the following example for an XML button created :
```
        <com.mercadolibre.android.andesui.button.AndesButton
            android:id="@id/button_id"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            app:andesButtonSize="large"
            app:andesButtonHierarchy="loud" />
```
Now, the button hierarchy can be updated on runtime by accessing its properties:
```
button_id.hierarchy = AndesButtonHierarchy.TRANSPARENT
```

## Affected component
`AndesButton`

## Frontify component link
[https://company-161429.frontify.com/d/kxHCRixezmfK/n-a#/componentes/button](https://company-161429.frontify.com/d/kxHCRixezmfK/n-a#/componentes/button)

## Screenshots
Screenshot_1580504249

## Checks
Are you sure you followed all those steps? In such case, let's say with me "I solemnly declare that ...":
- [X] I have coded an example inside the Demo App,
- [X] I have added proper tests,
- [X] I have modified the [Changelog](https://github.com/mercadolibre/...) file,
- [ ] I have updated GitHub wiki,
- [ ] I have the explicit approval of one or more members of the UX Team,
- [X] I have updated the version in gradle.properties file.

## Change type
This is easy, let us know what this code is about:
- [ ] This code adds a new component
- [X] This code includes improvements to an existent component
- [ ] This code improves or modifies ONLY the Demo App.

### CODE REVIEW

**Comment by furyGorgoryAdmin** [comment]
### Informative Menu <h2>  ### 

### Use any of these tags in a pull request description or leave a comment ### 
| Tag | Description |
|---|---|
| **[canary]**| Creates a build that can only be deployed with canary strategy. | 
| **[build]** |  Creates version or build from a branch that has been tagged as "no releasable". | 
| **[name:version-name]** |Specify the name of the build. For example: [build] [name:1.2.3-name] or [canary] [name:1.2.3-name].|
| **[major]** | To automatically increase the first position of the version. For example: [build] [major] or [canary] [major].|
| **[minor]** | To automatically increase the second position of the version. For example: [build] [minor] or [canary] [minor]. |
| **[patch]** | To automatically increase the third position of the version. For example: [build] [patch] or [canary] [patch].|  

[For more information about ci-cd](http://docs.furycloud.io/?lang=esp&section=businessassurance)
[For more information about semver](https://semver.org/#summary)

File: components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt [hunk 1]
```diff
@@ -48,35 +50,72 @@ import com.mercadolibre.android.andesui.button.size.AndesButtonSize
...
+ *
  */
 class AndesButton : ConstraintLayout {
 
-    internal lateinit var textComponent: TextView
+    private lateinit var andesButtonAttrs: AndesButtonAttrs
     internal lateinit var leftIconComponent: ImageView
     internal lateinit var rightIconComponent: ImageView
-    private lateinit var config: AndesButtonConfiguration
+    internal lateinit var textComponent: TextView
```
Comment (carrymaster) [inline components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt]: se accede desde fuera?

File: components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt [hunk 2]
```diff
@@ -48,35 +50,72 @@ import com.mercadolibre.android.andesui.button.size.AndesButtonSize
...
+        }
+
+    /**
+     * Getter and setter for [size].
+     */
+    var size: AndesButtonSize
+        get() = andesButtonAttrs.andesButtonSize
+        set(value) {
+            andesButtonAttrs = andesButtonAttrs.copy(andesButtonSize = value)
+            val config = createConfig()
```
Comment (carrymaster) [inline components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt]:
```
createConfig().also {
   setupHeight(it)
   .... 
}
```

File: components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt [hunk 3]
```diff
@@ -48,35 +50,72 @@ import com.mercadolibre.android.andesui.button.size.AndesButtonSize
...
+            updateDynamicComponents(config)
+            updateComponentsAlignment(config)
+        }
+
+    /**
+     * Simplest constructor for creating an AndesButton programmatically.
+     * Builds an AndesButton with Large Size and Hierarchy Loud by default.
+     */
+    constructor(context: Context) : super(context) {
+        initAttrs(SIZE_DEFAULT, HIERARCHY_DEFAULT, ICON_DEFAULT, TEXT_DEFAULT) //Consider AndesButtonIcon.NO_ICON
```
Comment (carrymaster) [inline components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt]: ese comment `//Consider` es necesario?
Si lo es no deberia ir en el JavaDoc?

File: components/src/main/java/com/mercadolibre/android/andesui/button/factory/AndesButtonAttrsParser.kt [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                                     val andesButtonSize: AndesButtonSize,
+                                     val andesButtonLeftIcon: Drawable?,
+                                     val andesButtonRightIcon: Drawable?,
+                                     val andesButtonText: String?,
+                                     val andesButtonEnabled: Boolean = true)
+
+/**
+ * This object parse the attribute set and return an instance of AndesButtonAttrs to be used by AndesButton
+ */
+
```
Comment (carrymaster) [inline components/src/main/java/com/mercadolibre/android/andesui/button/factory/AndesButtonAttrsParser.kt]: no se suelen dejar espacios en blanco entre el javadoc y la clase

File: components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt [hunk 5]
```diff
@@ -48,35 +50,72 @@ import com.mercadolibre.android.andesui.button.size.AndesButtonSize
...
-     * Constructor for creating an AndesButton programmatically with the specified [buttonSize], [buttonHierarchy] and no icon.
+     * Getter and setter for [hierarchy].
      */
-    constructor(context: Context, buttonSize: AndesButtonSize, buttonHierarchy: AndesButtonHierarchy) : super(context) {
-        initAttrs(buttonSize, buttonHierarchy, null)
+    var hierarchy: AndesButtonHierarchy
+        get() = andesButtonAttrs.andesButtonHierarchy
+        set(value) {
+            andesButtonAttrs = andesButtonAttrs.copy(andesButtonHierarchy = value)
+            val config = createConfig()
```
Comment (carrymaster) [inline components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt]:
```
createConfig().also {
   updateDynamicComponents(it)
   .... 
}
```

File: components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt (hunk 1 above)
Comment (lmadrazo) [reply components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt]: No, sólo se actualiza dentro de la clase, debería ser `private` tenes razón. Pero ante esto me surge el problema de no poder acceder a este recurso en el test (`AndesButonTest` file). Tenes algún approach para que pueda testearlo aunque esté declarado como privado? Gracias

File: components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt (hunk 3 above)
Comment (lmadrazo) [reply components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt]: No es necesario. Borrado.

File: components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt (hunk 1 above)
Comment (lmadrazo) [reply components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt]: @carrymaster Agregué el annotation `@VisibleForTesting` para cumplir esto que te comentaba y dejarlos como `private` a la vez

Comment (carrymaster) [reply components/src/main/java/com/mercadolibre/android/andesui/button/AndesButton.kt]: Si se usa en un test dejemoslo `internal`
No es lo mas elegante pero si lo mas facil de mantener.

**Pull request merged**